src/main/java/com/scottlogic/librarygradproject/BookService.java

```diff
@@ -0,0 +1,78 @@
+package com.scottlogic.librarygradproject;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.dao.EmptyResultDataAccessException;
+
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.List;
+import java.util.Optional;
+
+public class BookService {
+
+    private final BookRepository bookRepo;
+
+    @Autowired
+    public BookService(BookRepository bookRepo) {this.bookRepo = bookRepo;}
+
+    private Book validateBook(Book book) {
+        book.setIsbn(Optional.ofNullable(book.getIsbn()).orElse("").trim());
+        book.setAuthor(Optional.ofNullable(book.getAuthor()).orElseThrow(() -> new IncorrectBookFormatException()).trim());
+        book.setTitle(Optional.ofNullable(book.getTitle()).orElseThrow(() -> new IncorrectBookFormatException()).trim());
+        book.setPublishDate(Optional.ofNullable(book.getPublishDate()).orElse("").trim());
+        String year = book.getPublishDate();
+        if (year.length() > 0) {
+            while (year.length() < 4) {
+                year = "0" + year;
+            }
+        }
+        book.setPublishDate(year);
+        if (!book.getIsbn().matches("|[0-9]{10}|[0-9]{13}") ||
+                (book.getAuthor().length() == 0 || book.getAuthor().length() > 200) ||
+                (book.getTitle().length() == 0 || book.getTitle().length() > 200) ||
+                (book.getPublishDate().length() > 0 && (!book.getPublishDate().matches("|[0-9]{4}") ||
+                        Integer.parseInt(book.getPublishDate()) > Calendar.getInstance().get(Calendar.YEAR)))) {
+            throw new IncorrectBookFormatException();
+        }
+        return book;
+    }
+
+    public List<Book> findAll() {return bookRepo.findAll();}
+
+    public Book findOne(long id) {
+        Optional<Book> bookToGet = Optional.ofNullable(bookRepo.findOne(id));
+        return bookToGet.orElseThrow(() -> new BookNotFoundException(id));
+    }
+
+    public void delete(long id) {
+        try {
+            bookRepo.delete(id);
+        }
+        catch (EmptyResultDataAccessException e) {
+            throw new BookNotFoundException(id);
+        }
+    }
+
+    public void save(Book book) {
+        Book newBook = this.validateBook(book);
+        bookRepo.save(newBook);
+    }
+
+    public void put(Book book, long id) {
+        this.findOne(id);
```

Comment:
What happens to the return?

Reply:
we dont need it, the idea is that if the book is found we can continue, if its not findOne(id) will throw an exception with 404 code informing the user book with that id was not found.